src/julia_gc.c

```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkGapObjSafe(void * obj)
+{
+    // only traverse objects internally used by GAP
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
+        return 0;
+    return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
+}
+
 static inline int JMark(void * obj)
 {
+#ifdef VALIDATE_MARKING
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
+        abort();
+    if (!jl_typeis(ty, jl_datatype_type))
+        abort();
+#endif
     return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
 }
 
+void MarkJuliaObjSafe(void * obj)
+{
+    if (!obj)
+        return;
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
```

Comment:
So we call `jl_gc_internal_obj_base_ptr` on the type -- why don't we just call `jl_gc_internal_obj_base_ptr` on the object, i.e., `jl_gc_internal_obj_base_ptr(obj)`?

Reply:
Same question of course applies to the debug code in `JMark`

Reply:
At this point we already know that `obj` points to a proper object, which may be either a pool or large object. What the check does is verifying that it actually is a proper Julia object and has not been freed; if freed, the type tag would be a freelist pointer and fail that check.